File: client/src/main/java/au/com/codeka/warworlds/client/ui/SharedViews.java
package au.com.codeka.warworlds.client.ui;

import android.content.Context;
import android.view.View;

import com.google.common.base.Preconditions;

import java.util.ArrayList;

/**
 * Container class for holding shared views that we want to animate specially between screen
 * transitions.
 */
public class SharedViews {
  private final ArrayList<SharedView> sharedViews;

  private SharedViews(ArrayList<SharedView> sharedViews) {
    this.sharedViews = Preconditions.checkNotNull(sharedViews);
  }

  public ArrayList<SharedView> getSharedViews() {
    return sharedViews;
  }

  public String toDebugString(Context context) {
    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < sharedViews.size(); i++) {
      if (i == 0) {
        sb.append("[");
      } else {
        sb.append(", ");
      }
      sb.append(sharedViews.get(i).toDebugString(context));
    }
    sb.append("]");
    return sb.toString();
  }

  public static Builder builder() {
    return new Builder();
  }

  public static class SharedView {
    private final int viewId;
    private final View fromView;
    private final int fromViewId;
    private final int toViewId;

    public SharedView(int viewId) {
      this.viewId = viewId;
      this.fromView = null;
      this.fromViewId = 0;
      this.toViewId = 0;
    }

    public SharedView(int fromViewId, int toViewId) {
      this.viewId = 0;
      this.fromView = null;
      this.fromViewId = fromViewId;
      this.toViewId = toViewId;
    }

    public SharedView(View fromView, int toViewId) {
      this.viewId = 0;
      this.fromView = fromView;
      this.fromViewId = 0;
      this.toViewId = toViewId;
    }

    public int getViewId() {
      return viewId;
    }

    public View getFromView() {
      return fromView;
    }

    public int getFromViewId() {
      return fromViewId;
    }

    public int getToViewId() {
      return toViewId;
    }

    public String toDebugString(Context context) {
      if (viewId != 0) {
        return context.getResources().getResourceName(viewId);
      } else {
        String str;
        if (fromView != null) {
          str = fromView.toString();
        } else {
          str = context.getResources().getResourceName(fromViewId);
        }
        str += " -> ";
        str += context.getResources().getResourceName(toViewId);
        return str;
      }
    }
  }

  public static class Builder {
    private final ArrayList<SharedView> sharedViews = new ArrayList<>();

    public Builder addSharedView(int viewId) {
      sharedViews.add(new SharedView(viewId));
      return this;
    }

    public Builder addSharedView(int fromViewId, int toViewId) {
      sharedViews.add(new SharedView(fromViewId, toViewId));
      return this;
    }

    public Builder addSharedView(View fromView, int toViewId) {
      sharedViews.add(new SharedView(fromView, toViewId));
      return this;
    }

    public SharedViews build() {
      return new SharedViews(sharedViews);
    }
  }
}

File: client/src/main/java/au/com/codeka/warworlds/client/ui/SharedViews.kt
package au.com.codeka.warworlds.client.ui

import android.content.Context
import android.view.View
import com.google.common.base.Preconditions
import java.util.*

/**
 * Container class for holding shared views that we want to animate specially between screen
 * transitions.
 */
class SharedViews private constructor(sharedViews: ArrayList<SharedView>) {
  val sharedViews: ArrayList<SharedView>

  fun toDebugString(context: Context?): String {
    val sb = StringBuilder()
    for (i in sharedViews.indices) {
      if (i == 0) {
        sb.append("[")
      } else {
        sb.append(", ")
      }
      sb.append(sharedViews[i].toDebugString(context))
    }
    sb.append("]")
    return sb.toString()
  }

  class SharedView {
    val viewId: Int
    val fromView: View?
    val fromViewId: Int
    val toViewId: Int

    constructor(viewId: Int) {
      this.viewId = viewId
      fromView = null
      fromViewId = 0
      toViewId = 0
    }

    constructor(fromViewId: Int, toViewId: Int) {
      viewId = 0
      fromView = null
      this.fromViewId = fromViewId
      this.toViewId = toViewId
    }

    constructor(fromView: View?, toViewId: Int) {
      viewId = 0
      this.fromView = fromView
      fromViewId = 0
      this.toViewId = toViewId
    }

    fun toDebugString(context: Context?): String {
      return if (viewId != 0) {
        context!!.resources.getResourceName(viewId)
      } else {
        var str: String
        str = fromView?.toString() ?: context!!.resources.getResourceName(fromViewId)
        str += " -> "
        str += context!!.resources.getResourceName(toViewId)
        str
      }
    }
  }

  class Builder {
    private val sharedViews = ArrayList<SharedView>()
    fun addSharedView(viewId: Int): Builder {
      sharedViews.add(SharedView(viewId))
      return this
    }

    fun addSharedView(fromViewId: Int, toViewId: Int): Builder {
      sharedViews.add(SharedView(fromViewId, toViewId))
      return this
    }

    fun addSharedView(fromView: View?, toViewId: Int): Builder {
      sharedViews.add(SharedView(fromView, toViewId))
      return this
    }

    fun build(): SharedViews {
      return SharedViews(sharedViews)
    }
  }

  companion object {
    @JvmStatic
    fun builder(): Builder {
      return Builder()
    }
  }

  init {
    this.sharedViews = Preconditions.checkNotNull(sharedViews)
  }
}